demos/meltdown_ss.cc

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Demonstrates the Meltdown-SS. Speculative fetching of data from non-present
+ * segments and violating segment limits at the same time. Since segment limits
+ * are enforced only in 32-bit mode, this example does not work on x86_64.
+ * We initialize two segments - one points to public data, one points to
+ * private data. Then we use two segment pointers (FS and ES) and point them to
+ * those segments. Afterwards we make the segment pointing to private data
+ * non-present and limit its size to one character (which is dummy in the
+ * private data). Finally we speculatively read from the non-present segment
+ * beyond its size limits capturing the architectural failures with a SIGSEGV
+ * handler.
+ **/
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#ifndef __i386__
+#  error Unsupported architecture. 32-bit AMD required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <asm/ldt.h>
+#include <signal.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+// Have one dummy character on the beginning of the private data. Segment limit
+// zero means that exactly one address is accessible. We want to access only
+// non-accesible characters.
+const char *private_data = " It's a s3kr3t!!!";
+const char *public_data = "Hello, world!";
+
+// Sets up a segment descriptor in the local descriptor table.
+static void setup_segment(int index, const char *base, bool present) {
+  struct user_desc table_entry;
+  table_entry.entry_number = index;
+  table_entry.base_addr = reinterpret_cast<size_t>(base);
+  // No size limit for a present segment, one byte for a non-present segment.
+  table_entry.limit = 0xFFFFFFFF * static_cast<int>(present);
```

Comment:
```suggestion
  table_entry.limit = present ? 0xFFFFFFFF : 0;
```

I know we've gotten used to writing a bunch of branchless code, but no reason to avoid conditionals here. :)

Would also be useful to add to the comment to remind folks (as you did in the description at the top) that the `limit` is really the offset of the last accessible byte, so even a value of zero creates a one-byte segment.